demes/convert/ms.py

```diff
@@ -0,0 +1,757 @@
+        ),
+    )
+    group.add_argument(
+        "--migration-matrix-change",
+        "-ema",
+        nargs="+",
+        type=try_float,
+        default=[],
+        action=EventAction,
+        metavar="entry",
+        help=(
+            "Sets the migration matrix to the specified value at time t."
+            "The entries are in the order M[1, 1], M[1, 2], ..., M[2, 1],"
+            "M[2, 2], ..., M[N, N], where N is the number of populations."
+        ),
+    )
+
+    group = parser.add_argument_group("Demography")
+    group.add_argument(
+        "--growth-rate",
+        "-G",
+        metavar="alpha",
+        type=float,
+        help="Set the growth rate to alpha for all populations.",
+    )
+    group.add_argument(
+        "--population-growth-rate",
+        "-g",
+        action="append",
+        default=[],
+        nargs=2,
+        metavar=("population_id", "alpha"),
+        type=float,
+        help="Set the growth rate to alpha for a specific population.",
+    )
+    group.add_argument(
+        "--population-size",
+        "-n",
+        action="append",
+        default=[],
+        nargs=2,
+        metavar=("population_id", "size"),
+        type=float,
+        help="Set the size of a specific population to size*N0.",
+    )
+
+    group.add_argument(
+        "--growth-rate-change",
+        "-eG",
+        nargs=2,
+        action=EventAction,
+        type=float,
+        default=[],
+        metavar=("t", "alpha"),
+        help="Set the growth rate for all populations to alpha at time t",
+    )
+    group.add_argument(
+        "--population-growth-rate-change",
+        "-eg",
+        nargs=3,
+        action=EventAction,
+        type=float,
+        default=[],
+        metavar=("t", "population_id", "alpha"),
+        help=("Set the growth rate for a specific population to " "alpha at time t"),
+    )
+    group.add_argument(
+        "--size-change",
+        "-eN",
+        nargs=2,
+        action=EventAction,
+        type=float,
+        default=[],
+        metavar=("t", "x"),
+        help="Set the population size for all populations to x * N0 at time t",
+    )
+    group.add_argument(
+        "--population-size-change",
+        "-en",
+        nargs=3,
+        action=EventAction,
+        type=float,
+        default=[],
+        metavar=("t", "population_id", "x"),
+        help=(
+            "Set the population size for a specific population to " "x * N0 at time t"
+        ),
+    )
+    group.add_argument(
+        "--population-split",
+        "-ej",
+        nargs=3,
+        action=EventAction,
+        type=float,
+        default=[],
+        metavar=("t", "i", "j"),
+        help=(
+            "Move all lineages in population i to j at time t. "
+            "Forwards in time, this corresponds to a population split "
+            "in which lineages in j split into i. All migration "
+            "rates for population i are set to zero."
+        ),
+    )
```

Comment:
So, this `-ej` option has me stumped. Firstly, the statement here that "All migration rates for population i are set to zero" is wrong. The ms manual says
> -ej t i j
Move all lineages in subpopulation i to subpopulation j at
time t. Migration rates from subpopulation i (M[k,i], k ≠ i) are set to
zero. (Prior to May 19, 2007, ms left migration rates unchanged .) In
the forward direction this corresponds to population splitting. Population
growth rates are unchanged.

The change in 2007 is to turn off migration in one direction only. I checked the code, and this seems to be exactly what the code is doing (the ms code is fairly difficult to grok, so maybe I missed something). But I can't see why one would care about that direction. With the post-2007 behaviour, lineages are no longer allowed to migrate (backwards in time) from population `i`. But after the event, there won't be any lineages in that population anyway, so who cares? However, lineages *can* migrate (backwards in time) into population`i`, causing almost the same issue that plagued the Gutenkust et al. OOA model example in the msprime manual.

I didn't see anything in [Ragsdale et al. 2020](https://doi.org/10.1016/j.ajhg.2020.08.017) that suggested the original ms implementation of the Gutenkunst et al. OOA model suffered from this problem. So either (1) I'm wrong, or (2) the ms code change in 2007 is wrong, and the indices should be reversed.

Reply:
Perhaps foolishly, I assumed that the ms manual describes migration matrix indices and migration direction like msprime, where the source/dest of migration are with respect to lineages moving as the coalescent process progresses backwards through time. Indeed the ms manual states:

> ms will also produces samples under models with arbitrary backward mi-
gration matrices. This is done by first using the -I option and its arguments
(with or without the migration rate parameter.) This establishes the number
of subpopulations, the sample configuration, and assigns preliminary values to
the elements of the migration matrix. The elements of the migration matrix
are M[i][j] = 4*N0*m[i][j], i, j = 1, . . . ,npop , where m[i][j] is the fraction of subpopu-
lation i which is made up of migrants from subpopulation j each generation.

The first sentence suggests a backwards-time convention, but the last sentence could be interpreted as forwards-time wording. So is ms really is using forwards time conventions for the indices of `M` (and `m`)?

The 2007 fix makes no sense otherwise. Further, if the backwards convention applied, the Gutenkunst OOA model would potentially run forever (or give an "infinite coalescent time" error) when migrants entered an already merged population and could not escape (I tried running 100 reps with 300 chroms and it completed just fine). Finally, I dug further into the code. See below from `msdir/streec.c`, where `migm` is the migration matrix (and I guess the diagonal elements are temp variables, perhaps column or row sums, although I didn't figure that bit out yet).

```c
	    } else if (event == 'm') {	/* migration event */
		x = mig * ran1();
		sum = 0.0;
		for (i = 0; i < nchrom; i++) {
		    sum += migm[chrom[i].pop][chrom[i].pop];
		    if (x < sum)
			break;
		}
		migrant = i;
		x = ran1() * migm[chrom[i].pop][chrom[i].pop];
		sum = 0.0;
		for (i = 0; i < npop; i++) {
		    if (i != chrom[migrant].pop) {
			sum += migm[chrom[migrant].pop][i];
			if (x < sum)
			    break;
		    }
		}
		source_pop = i;
		config[chrom[migrant].pop] -= 1;
		config[source_pop] += 1;
		chrom[migrant].pop = source_pop;
	    }
```

Help?

Reply:
This is a bit tricky to wade through.  Let me know if you don't figure it out--I may be able to dust off my memories here.

Reply:
Thanks @molpopgen. I've woken up this morning with the idea that I should treat this as a black box and just construct a simple model demographic model to test it: if I guess the migration convention correctly, samples will coalesce, and if I guess incorrectly there will be an inifinite waiting time.

Reply:
ms is using the forwards-time semantics/language for all migration indices